esp-hal/src/spi/master.rs

```diff
@@ -658,14 +692,18 @@ where
     /// Assign the MOSI (Master Out Slave In) pin for the SPI instance.
     ///
     /// Enables output functionality for the pin, and connects it to the MOSI.
+    /// Calling this will replace previous pin assignments for this signal.
     ///
     /// You want to use this for full-duplex SPI or
     /// [DataMode::SingleTwoDataLines]
-    pub fn with_mosi<MOSI: PeripheralOutput>(self, mosi: impl Peripheral<P = MOSI> + 'd) -> Self {
+    pub fn with_mosi<MOSI: PeripheralOutput>(
+        mut self,
+        mosi: impl Peripheral<P = MOSI> + 'd,
+    ) -> Self {
         crate::into_mapped_ref!(mosi);
         mosi.enable_output(false);
 
-        self.driver().info.mosi.connect_to(&mut mosi);
+        self.pins.mosi_pin = OutputConnection::connect_with_guard(mosi, self.driver().info.mosi);
```

Comment:
I'm not quite happy that we need to pass in the peripheral signal here. :/ We could do `OutputConnection::connect_with_guard(mosi, &mut self.pins.mosi_pin)` instead, but whether that's better is another question.

Reply:
I'm not sure what we can do better here, either we store the signal or the pin directly into `State` right? Unless I'm missing something (which is likely).

If it's a case of storing the pin or signal, imo storing the signal is fine here.

Reply:
We don't want to store anything in State, since we don't drop State 😅 My issue is just that we have to pass the signal, which comes with a possibility of passing in the wrong signal. It's not a big deal I guess.